File: dashlane-project/Dashlane/src/main/java/com/dashlane/ui/credential/passwordgenerator/PasswordGeneratorDialog.java
package com.dashlane.ui.credential.passwordgenerator;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.StringDef;
import androidx.fragment.app.FragmentManager;

import com.dashlane.R;
import com.dashlane.dagger.singleton.SingletonProvider;
import com.dashlane.hermes.LogRepository;
import com.dashlane.passwordstrength.PasswordStrength;
import com.dashlane.ui.dialogs.fragments.NotificationDialogFragment;
import com.dashlane.ui.fragments.PasswordGenerationCallback;
import com.dashlane.ui.fragments.PasswordGeneratorFragment;
import com.dashlane.useractivity.log.inject.UserActivityComponent;
import com.dashlane.useractivity.log.usage.UsageLogRepository;
import com.dashlane.vault.model.VaultItem;
import com.dashlane.xml.domain.SyncObject;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;



@SuppressWarnings("java:S110")
public class PasswordGeneratorDialog extends NotificationDialogFragment implements PasswordGenerationCallback {

    @StringDef({Origin.EDIT_VIEW, Origin.CREATION_VIEW})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Origin {
        String EDIT_VIEW = "editPasswordView";
        String CREATION_VIEW = "addPasswordView";
    }

    @Origin
    private String mOrigin;
    private String mDomainAsking;
    private Callback mPasswordCallback;

    public interface Callback {
        void onPasswordAccepted(VaultItem<SyncObject.GeneratedPassword> generatedPassword);
    }

    @Override
    public void onPasswordGenerated() {
        
        setButtonEnable(DialogInterface.BUTTON_POSITIVE, true);
    }

    @Override
    public void passwordSaved(VaultItem<SyncObject.GeneratedPassword> generatedPassword, PasswordStrength strength) {
        getPasswordGeneratorLogger().log(mOrigin, "clickUse",
                (strength == null) ? null : String.valueOf(strength.getPercentScore()));
        if (mPasswordCallback != null) {
            mPasswordCallback.onPasswordAccepted(generatedPassword);
        }
    }

    @Override
    public void passwordGeneratedColor(int color) {
        
    }

    @Override
    public void restoreDominantColor(int color) {
        
    }

    @Override
    public void showPreviouslyGenerated() {
        
    }

    @Override
    public View onCreateDialogCustomView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_dialog_password_generator, container, false);
        ViewGroup fragmentViewGroup = v.findViewById(R.id.dialogPasswordGeneratorFragment);
        View copyButton = fragmentViewGroup.findViewById(R.id.copy_generated_password);
        copyButton.setVisibility(View.GONE);

        PasswordGeneratorFragment generatorFragment = getPasswordGeneratorFragment();
        if (generatorFragment != null) {
            generatorFragment.setLog75Subtype(mOrigin);
            generatorFragment.setPasswordGenerationCallback(this);
        }
        return v;
    }

    @Override
    public void show(FragmentManager manager, String tag) {
        super.show(manager, tag);
        getPasswordGeneratorLogger().log(mOrigin, "clickGenerate", null);
    }

    @Override
    public void onStart() {
        super.onStart();
        
        setButtonEnable(DialogInterface.BUTTON_POSITIVE, false);
    }

    @Override
    protected void onClickPositiveButton() {
        super.onClickPositiveButton();

        PasswordGeneratorFragment fragment = getPasswordGeneratorFragment();
        if (fragment != null) {
            fragment.copyAndSaveGeneratedPassword(mDomainAsking, false);
        }
    }

    @Override
    protected void onClickNegativeButton() {
        super.onClickNegativeButton();
        getPasswordGeneratorLogger().log(mOrigin, "clickCancel", null);
    }


    @Override
    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);
        PasswordGeneratorFragment fragment = getPasswordGeneratorFragment();
        if (fragment != null) {
            getActivity().getSupportFragmentManager().beginTransaction().remove(fragment).commitNowAllowingStateLoss();
        }
    }

    public void setOrigin(@Origin String origin) {
        mOrigin = origin;
    }

    public void setDomainAsking(String domainAsking) {
        mDomainAsking = domainAsking;
    }

    private PasswordGeneratorLogger getPasswordGeneratorLogger() {
        Context context = getContext();
        LogRepository trackingRepository = SingletonProvider.getComponent().getLogRepository();
        if (context == null) return new PasswordGeneratorLogger(null, trackingRepository);

        UsageLogRepository usageLogRepository =
                UserActivityComponent.Companion.invoke(context).getCurrentSessionUsageLogRepository();
        return new PasswordGeneratorLogger(usageLogRepository, trackingRepository);
    }

    PasswordGeneratorFragment getPasswordGeneratorFragment() {
        return (PasswordGeneratorFragment) getActivity().getSupportFragmentManager()
                                                        .findFragmentById(R.id.dialogPasswordGeneratorFragment);
    }

    public void setPasswordCallback(Callback callback) {
        mPasswordCallback = callback;
    }

    private static class StrengthNotComputedException extends Exception {
        private StrengthNotComputedException(String message) {
            super(message);
        }
    }
}

File: dashlane-project/Dashlane/src/main/java/com/dashlane/ui/credential/passwordgenerator/PasswordGeneratorDialog.kt
package com.dashlane.ui.credential.passwordgenerator

import android.app.Activity
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringDef
import androidx.core.os.bundleOf
import androidx.fragment.app.setFragmentResult
import com.dashlane.R
import com.dashlane.hermes.LogRepository
import com.dashlane.passwordstrength.PasswordStrength
import com.dashlane.session.SessionManager
import com.dashlane.ui.dialogs.fragments.NotificationDialogFragment
import com.dashlane.ui.fragments.PasswordGenerationCallback
import com.dashlane.ui.fragments.PasswordGeneratorFragment
import com.dashlane.vault.model.VaultItem
import com.dashlane.xml.domain.SyncObject
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import kotlinx.coroutines.ObsoleteCoroutinesApi

@OptIn(ObsoleteCoroutinesApi::class)
@AndroidEntryPoint
class PasswordGeneratorDialog : NotificationDialogFragment(), PasswordGenerationCallback {

    @Inject
    lateinit var logRepository: LogRepository

    @Inject
    lateinit var sessionManager: SessionManager

    @Origin
    lateinit var origin: String
    private lateinit var domainAsking: String

    private val passwordGeneratorFragment: PasswordGeneratorFragment
        get() = requireActivity().supportFragmentManager
            .findFragmentById(R.id.dialogPasswordGeneratorFragment) as PasswordGeneratorFragment

    override fun onPasswordGenerated() {
        
        setButtonEnable(DialogInterface.BUTTON_POSITIVE, true)
    }

    override fun passwordSaved(
        generatedPassword: VaultItem<SyncObject.GeneratedPassword>,
        strength: PasswordStrength?
    ) {
        setFragmentResult(
            PASSWORD_GENERATOR_REQUEST_KEY,
            bundleOf(
                PASSWORD_GENERATOR_RESULT_ID to generatedPassword.syncObject.id,
                PASSWORD_GENERATOR_RESULT_PASSWORD to generatedPassword.syncObject.password.toString()
            )
        )
    }

    override fun restoreDominantColor(color: Int) {
        
    }

    override fun showPreviouslyGenerated() {
        
    }

    override fun onCreateDialogCustomView(
        inflater: LayoutInflater?,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater?.inflate(R.layout.fragment_dialog_password_generator, container, false)
        view?.apply {
            val fragmentViewGroup = findViewById<ViewGroup>(R.id.dialogPasswordGeneratorFragment)
            val copyButton = fragmentViewGroup.findViewById<View>(R.id.copy_generated_password)
            copyButton.visibility = View.GONE
        }
        passwordGeneratorFragment.log75Subtype = origin
        passwordGeneratorFragment.passwordGenerationCallback = this
        return view
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        origin = requireArguments().getString(PASSWORD_GENERATOR_ARG_ORIGIN)!!
        domainAsking = requireArguments().getString(PASSWORD_GENERATOR_ARG_DOMAIN)!!

        
        setButtonEnable(DialogInterface.BUTTON_POSITIVE, false)
    }

    override fun onClickPositiveButton() {
        super.onClickPositiveButton()
        passwordGeneratorFragment.copyAndSaveGeneratedPassword(domainAsking, false)
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        requireActivity().supportFragmentManager.beginTransaction().remove(passwordGeneratorFragment)
            .commitNowAllowingStateLoss()
    }

    @StringDef(EDIT_VIEW, CREATION_VIEW)
    @Retention(AnnotationRetention.SOURCE)
    annotation class Origin

    companion object {
        const val DIALOG_PASSWORD_GENERATOR_TAG = "PASSWORD_GENERATOR_POPUP"
        const val PASSWORD_GENERATOR_ARG_ORIGIN = "password_generator_arg_origin"
        const val PASSWORD_GENERATOR_ARG_DOMAIN = "password_generator_arg_domain"
        const val PASSWORD_GENERATOR_REQUEST_KEY = "password_generator_request_key"
        const val PASSWORD_GENERATOR_RESULT_ID = "password_generator_result_id"
        const val PASSWORD_GENERATOR_RESULT_PASSWORD = "password_generator_result_password"
        const val EDIT_VIEW = "editPasswordView"
        const val CREATION_VIEW = "addPasswordView"

        fun newInstance(
            activity: Activity,
            @Origin origin: String,
            domainAsking: String
        ): PasswordGeneratorDialog {
            return Builder()
                .setArgs(
                    bundleOf(
                        PASSWORD_GENERATOR_ARG_ORIGIN to origin,
                        PASSWORD_GENERATOR_ARG_DOMAIN to domainAsking
                    )
                )
                .setPositiveButtonText(activity, R.string.use)
                .setNegativeButtonText(activity, R.string.cancel)
                .setClickNegativeOnCancel(true)
                .build(PasswordGeneratorDialog())
        }
    }
}